crates/stark/src/machine.rs

```diff
@@ -185,30 +192,54 @@ impl<SC: StarkGenericConfig, A: MachineAir<Val<SC>>> StarkMachine<SC, A> {
         initial_global_cumulative_sum: SepticDigest<Val<SC>>,
     ) -> (StarkProvingKey<SC>, StarkVerifyingKey<SC>) {
         let parent_span = tracing::debug_span!("generate preprocessed traces");
-        let mut named_preprocessed_traces = parent_span.in_scope(|| {
-            self.chips()
-                .par_iter()
-                .filter_map(|chip| {
-                    let chip_name = chip.name();
-                    let begin = Instant::now();
-                    let prep_trace = chip.generate_preprocessed_trace(program);
-                    tracing::debug!(
-                        parent: &parent_span,
-                        "generated preprocessed trace for chip {} in {:?}",
-                        chip_name,
-                        begin.elapsed()
-                    );
-                    // Assert that the chip width data is correct.
-                    let expected_width = prep_trace.as_ref().map_or(0, p3_matrix::Matrix::width);
-                    assert_eq!(
-                        expected_width,
-                        chip.preprocessed_width(),
-                        "Incorrect number of preprocessed columns for chip {chip_name}"
-                    );
-                    prep_trace.map(move |t| (chip_name, chip.local_only(), t))
-                })
-                .collect::<Vec<_>>()
-        });
+        let (named_preprocessed_traces, num_constraints): (Vec<_>, Vec<_>) =
+            parent_span.in_scope(|| {
+                self.chips()
+                    .par_iter()
+                    .map(|chip| {
+                        let chip_name = chip.name();
+                        let begin = Instant::now();
+                        let prep_trace = chip.generate_preprocessed_trace(program);
+                        tracing::debug!(
+                            parent: &parent_span,
+                            "generated preprocessed trace for chip {} in {:?}",
+                            chip_name,
+                            begin.elapsed()
+                        );
+                        // Assert that the chip width data is correct.
+                        let expected_width =
+                            prep_trace.as_ref().map_or(0, p3_matrix::Matrix::width);
+                        assert_eq!(
+                            expected_width,
+                            chip.preprocessed_width(),
+                            "Incorrect number of preprocessed columns for chip {chip_name}"
+                        );
+
+                        // Count the number of constraints.
+                        let num_constraints = get_symbolic_constraints(
+                            &chip.air,
+                            chip.preprocessed_width(),
+                            PROOF_MAX_NUM_PVS,
+                        )
+                        .len();
+
+                        let permutation_constraints = count_permutation_constraints(
```

Comment:
nit: Consider renaming to `num_permutation_constraints` to make naming a bit more consistent.